src/tiled/scriptpropertytype.h

```diff
@@ -0,0 +1,131 @@
+#include <QObject>
+
+namespace Tiled {
+/**
+ * Scripting engine wrapper for PropertyType
+ */
+class ScriptPropertyType : public QObject
+{
+    Q_OBJECT
+    Q_PROPERTY(QString name READ name)
+    Q_PROPERTY(bool isClass READ isClass)
+    Q_PROPERTY(bool isEnum READ isEnum)
+    Q_PROPERTY(QVariant defaultValue READ defaultValue)
+
+public:
+    ScriptPropertyType(const SharedPropertyType &propertyType)
+        : mType(propertyType)
+    {}
+
+    const QString &name() const;
+    bool isClass() const { return mType->isClass(); }
+    bool isEnum() const { return mType->isEnum(); }
+    QVariant defaultValue() { return mType->defaultValue(); }
+
+private:
+    SharedPropertyType mType;
+};
+
+class ScriptEnumPropertyType : public ScriptPropertyType
+{
+    Q_OBJECT
+
+    Q_PROPERTY(StorageType storageType READ storageType)
+    Q_PROPERTY(QStringList values READ values)
+
+public:
+    ScriptEnumPropertyType(const QSharedPointer<EnumPropertyType> &propertyType)
+        : ScriptPropertyType(propertyType)
+        , mEnumType(propertyType)
+    {}
+    // copied from propertytype.h
+    enum StorageType {
+        StringValue,
+        IntValue
+    };
+    Q_ENUM(StorageType);
+
+    StorageType storageType() const { return static_cast<StorageType>(mEnumType->storageType); }
+    QStringList values() const { return mEnumType->values; }
+
+private:
+    QSharedPointer<EnumPropertyType> mEnumType;
+};
+
+class ScriptClassPropertyType : public ScriptPropertyType
+{
+    Q_OBJECT
+    Q_PROPERTY(QColor color READ color)
+    Q_PROPERTY(QVariantMap members READ members)
+    Q_PROPERTY(bool drawFill READ drawFill)
+    Q_PROPERTY(int usageFlags READ usageFlags)
+
+public:
+    ScriptClassPropertyType(const QSharedPointer<ClassPropertyType> &propertyType)
+        : ScriptPropertyType(propertyType)
+        , mClassType(propertyType)
+    {}
+
+    // TODO: a way to avoid duplicating this again? 
+    enum ClassUsageFlag {
+        PropertyValueType   = 0x001,
+
+        // Keep values synchronized with Object::TypeId
+        LayerClass          = 0x002,
+        MapObjectClass      = 0x004,
+        MapClass            = 0x008,
+        TilesetClass        = 0x010,
+        TileClass           = 0x020,
+        WangSetClass        = 0x040,
+        WangColorClass      = 0x080,
+        ProjectClass        = 0x100,
+        AnyUsage            = 0xFFF,
+        AnyObjectClass      = AnyUsage & ~PropertyValueType,
+    };
+    Q_ENUM(ClassUsageFlag)
+
+    QColor color() const { return mClassType->color; }
+    // TODO: " No viable overloaded '=' "
+    // void setColor(QColor &value) { mClassType->color = value; }
+    QVariantMap members() const {return mClassType->members; }
+    bool drawFill() const { return mClassType->drawFill; }
+    // void setDrawFill(bool value) { mClassType->drawFill = value; }
+    int usageFlags() const { return mClassType->usageFlags; }
+    //void setUsageFlags(int value) { mClassType->setUsageFlags(value); }
```

Comment:
The setter for usage flags accepts `bool value`  to either turn the specified flags on or off, but it doesn't in itself  provide a way to just set the usage flags to the specified value. Would script writers expect to be able to just hard set the usage flags to a value (personally, this is what I think makes sense so that you can do things like &= , |= the flags)  or have a method like setUsageFlags(flags,  value: bool) ?  

Without changing the setUsageFlags in propertytype.cpp, to achieve a setter, I could turn off all flags with `setUsageFlags( ClassUsageFlag. AnyUsage  , false)` and then turn the specified flags on with  `setUsageFlags(flagsFromScriptUser, true) ` if that makes sense

Reply:
Indeed that signature is quite UI-oriented since it's used in response to the user toggling checkboxes. I agree from a scripting API point of view, we probably just want to be able to set the flags directly and you could introduce a `void ClassPropertyType::setUsageFlags(int flags)` overload for this.